### ISSUE TITLE
General Retrospective for March/April 2024 Releases

### ISSUE THREAD

**Issue by adamfarley** [adoptium/temurin#28]
**Summary**

A retrospective for all efforts surrounding the titular releases. 

All community members are welcome to contribute to the agenda via comments below.

This will be a virtual meeting after the release, with at least a week of notice in the #release Slack channel.

On the day of the meeting we'll review the agenda and add a list of actions at the end.

Invited: Everyone.

**Time, Date, and URL**

Time: 3pm BST, 10am EST.
Date: Tuesday the 7th of May, 2024.
URL: https://eclipse.zoom.us/j/82423919203?pwd=jcs9cimNWYIflSqChjnT5U5Aj62sSx.1
Meeting ID: 824 2391 9203
Passcode: 339984

**Details**

Retrospective Owner Tasks (in order):

- [x] Post retro URL in \#Release around the start of the new release.
- [x] Wait until most builds are released, with no signs of a respin.
- [x] Announce the retrospective's date + time on \#Release a week in advance.
- [x] Tell Carmen Delgado so they can add this retrospective to the community calendar.
- [x] Host the retrospective:
  - Go through the agenda.
  - Create a list of actions.
- [ ] Process each action:
  - Create a "WIP" issue including the source comment.
  - Add the issue to the current iteration.
  - Add an issue link to the action list.
- [ ] Create a new retrospective issue for the next release.
- [ ] Set a calendar reminder so you remember to do step 1 before the next release.
- [ ] Close this issue.

**TLDR**

Add proposed agenda items as comments below.

**Comment by smlambert** [comment]
Scary msg when publishing aarch64_mac binaries, believe it did the right thing (pushing 31 artifacts to releases repo, but found 62 and reported UNSTABLE 
https://ci.adoptium.net/job/build-scripts/job/release/job/refactor_openjdk_release_tool/8424/

dryrun did not indicate any issues to be aware of https://ci.adoptium.net/job/build-scripts/job/release/job/refactor_openjdk_release_tool/8423/

**Comment by sophia-guo** [comment]
https://github.com/adoptium/temurin/issues/28#issuecomment-2010423578

The status is unstable as it's counting some of the -ea tagged artifacts, which means [the check file](https://github.com/adoptium/github-release-scripts/blob/master/sbin/releaseCheck.sh) needs to update. Maybe we should think is there a better way or automate way to check the numbers?

**Comment by jerboaa** [comment]
Something went wrong with publishing source tarballs for the Jan 2024 update. See: https://github.com/adoptium/adoptium-support/issues/1003 we should make sure that it's there with some verification for any release.

**Comment by sophia-guo** [comment]
Also to publish the binary you can try the rerun link at the lower part of page https://ci.adoptium.net/job/build-scripts/job/release-openjdk22-pipeline/5/, which is enabled for this release. 

Dry runs are triggered by pipeline job itself. If dry run fails the rerun link would be **Dry run RELEASE Publish temurin jdk-22+36 mac x64**, which will trigger a dryrun. Otherwise the rerun link would be **RELEASE Publish temurin jdk-22+36 mac x64** and no need to do the dryrun.
<img width="686" alt="Screenshot 2024-03-20 at 3 44 29 PM" src="https://github.com/adoptium/temurin/assets/18030625/fb135949-e17b-48ef-abc4-d3aeae0160a9">

**Comment by smlambert** [comment]
> Also to publish the binary you can try the rerun link 

That is what I clicked to run first the dry run (8423), then the release run (8424).  I expected the release checks that we have in place to work, but I guess they do not take into account the presence of the EA artifacts.

**Comment by smlambert** [comment]
With new feature release, check if Version List on website needs updating?  https://github.com/adoptium/adoptium.net/issues/2731

**Comment by sophia-guo** [comment]
> That is what I clicked to run first the dry run (8423), then the release run (8424).

A little bit of confused. Dry run was triggered by pipeline and succeeded. Is 8423 for double check?

**Comment by smlambert** [comment]
> A little bit of confused. Dry run was triggered by pipeline and succeeded. Is 8423 for double check? 

Yes, and also because I did not find the output from the dry run quickly

**Comment by sophia-guo** [comment]
Would it be helpful to add the dry-run build links?

**Comment by smlambert** [comment]
Downside of everyone and their dog using Grinders for all kinds of good work is that it is more difficult to spot the ones launched to complete release triage.  (no action required on this comment, good bookkeeping can cover it, or we advise folks doing dev work to use Grinder_Dev, or some such).

**Comment by smlambert** [comment]
> Would it be helpful to add the dry-run build links?

I think the dryruns for publishing were there to help deal with the great potential for human error, which is essentially removed by the addition of the prepopulated 'quick links' found at the bottom of the parent pipeline job.

If the dryruns did some of the verification that happens during the actual publish (checking the right number of artifacts exist), and subsequently fail or report if there was an issue, they would have a purpose.  Without the verification checks happening in the dry run, there seems little need to run an automated dry run.

**Comment by smlambert** [comment]
The release has been chugging along smoothly enough to allow for development work to continue alongside of it.  The unfreezing of master branches in Temurin project also a good thing.

**Comment by smlambert** [comment]
Triggering the pipeline ahead of the -ga tag was a good call.  The upstream -ga tag did not show up until much later (~full day) than we triggered the pipeline (Wed/20th).

**Comment by smlambert** [comment]
https://github.com/adoptium/aqa-tests/issues/5156#issuecomment-2025508317 - for follow-up AQAvit actions

**Comment by smlambert** [comment]
Release notes not being served up via API (so not showing up on the website).  [Slack msg](https://adoptium.slack.com/archives/CLCFNV2JG/p1711631642619729), am I missing a step that I do not find instructions for?

**Comment by smlambert** [comment]
Why is EclipseMirror job on the TC Jenkins server?  
Think we can hook that checklist action into a post build job, along with many other tasks.  (related: https://github.com/adoptium/ci-jenkins-pipelines/issues/610#issuecomment-2027151112)

**Comment by smlambert** [comment]
If we are clearing out / deleting jobs on Jenkins, let's do it via Jenkins API (which then keeps the JobID history), not by deleting workspace directly on the server (where JobID history not kept and ID count starts again, causing repeated/duplicated IDs).  This then leads to a problem with seeing the new runs on TRSS that uses those jobIDs as indices in the DB.

**Comment by sxa** [comment]
> Why is EclipseMirror job on the TC Jenkins server? Think we can hook that checklist action into a post build job, along with many other tasks. (related: [adoptium/ci-jenkins-pipelines#610 (comment)](https://github.com/adoptium/ci-jenkins-pipelines/issues/610#issuecomment-2027151112))

Because it contains a secret credential that we were only allowed to have on the eclipse-managed server.

**Comment by sxa** [comment]
> If we are clearing out / deleting jobs on Jenkins, let's do it via Jenkins API (which then keeps the JobID history), not by deleting workspace directly on the server (where JobID history not kept and ID count starts again, causing repeated/duplicated IDs). This then leads to a problem with seeing the new runs on TRSS that uses those jobIDs as indices in the DB.

See comment on https://github.com/adoptium/aqa-test-tools/issues/860#issuecomment-2034589098 - I don't believe that deleting the job definition via any means would have met the requirements here and retained those identifiers .

**Comment by sxa** [comment]
Guide to creating new mirror releases should including archiving the non-u release when u mirror is created.

**Comment by sxa** [comment]
The steps in https://github.com/adoptium/temurin-build/wiki/Creating-new-jdkNNu-(updates)-repro-mirror-from-the-jdkNN-release-mirror#how for manually doing initial population of the mirror should not be required based on [this slack thread](https://adoptium.slack.com/archives/C09NW3L2J/p1712317757749809?thread_ts=1712249567.967659&cid=C09NW3L2J) so the docs should reflet that (and possibly put the manual steps in a `<summary>` section. Noting also that after the mirror is created it can take up to three hours for the permissions to start working if you need to do it manually.
Also as per [the thread](https://adoptium.slack.com/archives/C09NW3L2J/p1712319605724919?thread_ts=1712249567.967659&cid=C09NW3L2J) you may see this error the first time you run a new mirror job to populate the repository which will cause the job to fail - it seems to go through ok on a second attempt to run the mirror job:
```
+ git rebase skara/master master
fatal: no such branch/commit 'master'
Build step 'Execute shell' marked build as failure
```

**Comment by sxa** [comment]
Also the doc on creating the generator shoulkd explicitly state that while the title of the job should have the `u` as appropriate, the `JAVA_VERSION` should NOT have the `u` as it gets added later and will result in this:
`hudson.remoting.ProxyException: java.nio.file.NoSuchFileException: /home/jenkins/workspace/build-scripts/utils/evaluation-pipeline_jobs_generator_jdk22u/pipelines/jobs/configurations/jdk22uu_pipeline_config.groovy`

**Comment by sxa** [comment]
Add note to the checklist or RELEASING.md to indicate that we generall use the same AQA branch for March+April and for Sept+Oct *Subject to updating the JDxx_BRANCH name for the "new" March/Sept release)

**Comment by sxa** [comment]
Noting that on my initial attempt to set the dryrun tags I made it `jdk-17.0.11+7-dryrun-ga` (i.e. including the `+7` build identifier. This will cause problems and such a tag will need to be deleted prior to running the pipelines or you'll get something like this from `openjdk_pipeline.groovy`:
```
[INFO] Resolved jdk-17.0.11-dryrun-ga to upstream build tag jdk-17.0.11+6jdk-17.0.11+6-dryrun-ga
[Pipeline] echo
[ERROR] scmReference does not match with any JDK branch in testenv.properties in aqa-tests release branch. Please update aqa-tests v1.0.1-release release branch. Set the current build result to FAILURE!
```
Also noting that if the pipeline does fail after being triggered, the `workspace/tracking` file on the jenkins worker node will need to be manually updated or you won't be able to re-trigger as the job uses that for its status and will not re-trigger the same underlying tag twice unless it's manually fixed.

**Comment by sxa** [comment]
The new `u` release (jdk22u) does not have any tags which can be used as an equivalent of the `-dryrun-ga` release, therefore we need to put in a fix to allow the dryrun process to run on jdk22u (and the same for subsequent STS releases). At the moment I'm going to use 20.0.0+0 because that will never be used, but we should consider what to do for future versions, since it will not be as simple to insert somethign similar between 20.0.1.x and 20.0.2.y
Ref: https://github.com/adoptium/mirror-scripts/pull/50

**Comment by sxa** [comment]
It seems that the code behind the `*-openjdk22-pipeilne` groovy scripts [prefers to pick up a non-`u` configuration](https://github.com/adoptium/ci-jenkins-pipelines/blob/a20199e9062526d06e7e5b782d851e94437f9765/pipelines/build/openjdk_pipeline.groovy#L180). I added 22u in [the configurations dir](https://github.com/adoptium/ci-jenkins-pipelines/tree/master/pipelines/jobs/configurations) and hadn't removed 22, so the dryruns triggered the `jdk22-` jobs using the `jdk22` repository instead of the right one.
Fixed by remmoving the jdk22 files from the configurations dir in [this PR](https://github.com/adoptium/ci-jenkins-pipelines/pull/996) but we should consider whether prefering the non-`u` version is correct (It's hard to envision a scenario where it would be IMHO) but we should definitely cover this in [the wiki page](https://github.com/adoptium/temurin-build/wiki/Setting-up-new-build-jobs,-mirrors,...-when-a-new-openjdk-version-is-branched-from-jdk(head)#create-pipeline_jobs_generator_jdknnu-job-and-configuration) to ensure that the new `u` releases are done with a rename instead of creating new ones alongside the non-`u` versions.

**Comment by sxa** [comment]
Should we use `vYYYY.MM.NN` or `vYYYY.MM+NN` for the build branches? The Releaseing guide is ambiguous (which resulted in me I ended up with both at one point):
![image](https://github.com/adoptium/temurin/assets/6487691/e5df9e8c-2afe-4ce8-8843-48f70c7f22a0)

**Comment by sxa** [comment]
Code freeze template for slack needs to be adjusted to indicate that not all build repositories are affected by it.

**Comment by sxa** [comment]
We should look at the criteria we use for disabling tests on the automated regular runs to ensure they do not cause problems with the dry runs: https://github.com/adoptium/ci-jenkins-pipelines/issues/1007

(Noting that this cycle was a bit special as I was trying to run all five releases in parallel to check for capacity and timings)

**Comment by sxa** [comment]
The [sign_verification](https://ci.adoptium.net/job/build-scripts/job/release/job/sign_verification) job is set to run on a test machine, and can therefore incur delays during a released cycle while such a machine becomes available (could be a while if it's the first released version and it gets queued up behind test jobs for other versions) There are delays of 4-5 hours for some jobs at the time of writing.

**Comment by sxa** [comment]
The release checklist has sections for the JDK8, 11 and "XX" releases - we should explicitly add 17 and 21 LTS releases in there too.

**Comment by andrew-m-leonard** [comment]
> The [sign_verification](https://ci.adoptium.net/job/build-scripts/job/release/job/sign_verification) job is set to run on a test machine, and can therefore incur delays during a released cycle while such a machine becomes available (could be a while if it's the first released version and it gets queued up behind test jobs for other versions) There are delays of 4-5 hours for some jobs at the time of writing. This job also seems to run outside a pipeline "block" in the jdkXX--pipeline job so it may not have been too obvious previously that it was holding things up.

We should move the signVerify() to before the kicking off of the aqaTests, so it is done first

**Comment by sxa** [comment]
The checklist currently has this:

> _Calculate the "expected" openjdk build tags for the releases being published, and update all the JDKnn_BRANCH values in the testenv.properties_ 

This will need to be updated in light of the updates that allow the correct tag to be autodetected. At present you do still need to have it set to the dryrun-ga ones for the dry-run process, but then switch to `-ga` for release time.

**Comment by sxa** [comment]
Reference to Setup_JCK_Multinode in the checklist is incorrect. Presumably should be Setup_JCK_Run_Multinode

**Comment by sxa** [comment]
We should clean up the old git-hg jobs from https://ci.adoptium.net/job/git-mirrors/ as they are only likely to cause confusion now. Do we even need the [adoptium](https://ci.adoptium.net/job/git-mirrors/job/adoptium/) subfolder?

**Comment by sxa** [comment]
Add item to process to ensure that the next release cycle is listed in the [adoptium calendar](https://calendar.google.com/calendar/u/0/embed?src=c_56d7263c0ceda87a1678f6144426f23fb53721480b5ff71b073afb51091e5492@group.calendar.google.com) so that EF are aware

**Comment by andrew-m-leonard** [comment]
New ready done Publish link very useful

**Comment by sophia-guo** [comment]
No Orka triggered with label [hw.arch.x86&&(sw.os.osx||sw.os.mac)&&ci.role.perf&&!sw.os.osx.10_14](https://ci.adoptium.net/label/hw.arch.x86&&%28sw.os.osx%7C%7Csw.os.mac%29&&ci.role.perf&&!sw.os.osx.10_14), which is required for perf jdk22.

[[hw.arch.x86&&(sw.os.osx||sw.os.mac)&&ci.role.perf&&!sw.os.osx.10_14](https://ci.adoptium.net/label/hw.arch.x86&&%28sw.os.osx%7C%7Csw.os.mac%29&&ci.role.perf&&!sw.os.osx.10_14)](https://github.com/adoptium/aqa-tests/issues/5235#issuecomment-2065614940)

Passed on 13th  and before on an orka system. No orka available for this release.

**Comment by smlambert** [comment]
Is there a naming convention that one should follow for build branches ?

<img width="324" alt="Screenshot 2024-04-21 at 5 52 43 PM" src="https://github.com/adoptium/temurin/assets/2836948/c4422c4b-cd60-4889-aeb3-223cd48771a8">

**Comment by sxa** [comment]
@smlambert Naming is covered in a [previous comment](https://github.com/adoptium/temurin/issues/28#issuecomment-2042897287) for discussion as the doc is currently ambiguous

**Comment by sxa** [comment]
Notes on blog post production:
- Should we be creating an issue or PR for the "next" release? (Checklist says PR, but we created an issue for this one)
- We should have a template for the blog post (maybe a script to set all the version numbers in the right place?) which includes information on how to generate the information for the vulnerability section (automatable?) and templates for how to display CA Certs updates too

**Comment by sxa** [comment]
Ensure we define processes for the installers for:
- Adding new distributions.
- Updating the adoptium-ca-certificates `.deb` package when required
Relevant issues/PRs for reference:
- https://github.com/adoptium/installer/issues/852
- https://github.com/adoptium/installer/pull/855
- https://github.com/adoptium/installer/pull/868

**Comment by smlambert** [comment]
> Should we be creating an issue or PR for the "next" release? (Checklist says PR, but we created an issue for this one)

I originally put PR, but think we should change it to issue, as it does what we need it to do (be a placeholder for 'new and noteworthy' notes between release period, and also does not tie the originator of the PR into being the next blog post author (pen named PMC).

**Comment by sxa** [comment]
> I originally put PR, but think we should change it to issue

Noting that I'm planning to avoid creating either just now until we do the retrospective and have the discussion on this (I'm writing this to remind me that I need to do it when we're going through these comments ;-) )

**Comment by sxa** [comment]
"Full update" had to be forced to pick up the new release notes ([slack thread](https://adoptium.slack.com/archives/CLCFNV2JG/p1714047634362939?thread_ts=1714033031.045209&cid=CLCFNV2JG)). Do we know why? Anythign we can fix / document for future understanding.

**Comment by sxa** [comment]
Release notes seem to [repeatedly have problems staging](https://github.com/adoptium/adoptium.net/issues/2808#issuecomment-2077638374). We should attempt to understand and resolve why.

**Comment by sxa** [comment]
download verification test gets confused if there aren't any .zip files in the release e.g. 22.0.1.1 for s390x: https://ci.adoptium.net/job/build-scripts/job/release/job/download_and_sbom_validation/27

**Comment by sxa** [comment]
Noting that choosing to run the initial pipelines without win32 meant that the 32-bit build+AQA pipelines completed in a maximum of [25 hours](https://github.com/adoptium/infrastructure/issues/3517#issuecomment-2088159986). This contrasts with the [initial tests](https://github.com/adoptium/infrastructure/issues/3517#issuecomment-2049833100) with both variants with all five releases which were taking up to 1d16h to complete. On this basis I would propose that we kick off the 32-bit pipelines approximately 24 hours after the win64 ones where practical and to avoid contention, although if all 64-bit pipelines are already in progress, then a short delay of 2 hours may be adequate (Bear in mind the goal is to avoid the 32-bit tests running first - the builds take between 26m (JDK8) and 95 minutes (JDK22) so a 2 hours buffer after the last 64-bit build will generally be adequate, although the 24 hour proposal means there there is less chance of machine contention for any re-runs for 64-bit re-runs that are required.

**Comment by sxa** [comment]
New feature: Improve download test capability so that it works with `ea` builds so we can pre-empt check issues during the release: https://github.com/adoptium/temurin-build/issues/3784

Potentially look at adding this to the main pipelines (Do we have a separate issue for that?)

**Comment by jiekang** [comment]
To discuss: https://github.com/adoptium/ci-jenkins-pipelines/issues/994

Priority: needs to be done before July release

**Comment by adamfarley** [comment]
**Actions**

https://github.com/adoptium/temurin/issues/28#issuecomment-2010483145
Add to download verification process - raise issue

https://github.com/adoptium/temurin/issues/28#issuecomment-2010612290
Raise issue for sxa to update release checklist

https://github.com/adoptium/temurin/issues/28#issuecomment-2013303022
Raise issue to review queue times in the next release.

https://github.com/adoptium/temurin/issues/28#issuecomment-2041364554
Create issue to discuss dryrun-launching job. Should we have one? Discuss.

https://github.com/adoptium/temurin/issues/28#issuecomment-2042703926
Discuss in cc.

https://github.com/adoptium/temurin/issues/28#issuecomment-2042865717
- discussion issue: do we need the u's in job names?
- Prioritise non-u configs over u configs? We shouldn't. Change code here & assign to Stewarts: https://github.com/adoptium/ci-jenkins-pipelines/blob/a20199e9062526d06e7e5b782d851e94437f9765/pipelines/build/openjdk_pipeline.groovy#L180

https://github.com/adoptium/temurin/issues/28#issuecomment-2051548415
Needs moving to same section as smoke tests (so it gets executed before the main test).

https://github.com/adoptium/temurin/issues/28#issuecomment-2057145466
create issue.

**Notes (any actions already handled as indicated)**

https://github.com/adoptium/temurin/issues/28#issuecomment-2010531753
Shelley has offered to raise an issue.

https://github.com/adoptium/temurin/issues/28#issuecomment-2027168432
Shelley to raise an aqa tests tools issue for this.

https://github.com/adoptium/temurin/issues/28#issuecomment-2041364554
Stewart to do doc change.

https://github.com/adoptium/temurin/issues/28#issuecomment-2049523037
Stewart to do this. No additional issue needed.

https://github.com/adoptium/temurin/issues/28#issuecomment-2051291870
Stewart to do this, no issue needed. Andrew to raise PR.

https://github.com/adoptium/temurin/issues/28#issuecomment-2051730654
Stewart to do, no issue needed.

https://github.com/adoptium/temurin/issues/28#issuecomment-2056461323
Issue already open, no issue needed.

**Progress**

The above actions are based on a partial review of the retrospective comments.

Resume this retrospective from comment https://github.com/adoptium/temurin/issues/28#issuecomment-2066290051
